### PR TITLE
Sketcher: Dimension: Fix 2 bugs.

### PR DESCRIPTION
**Pull request by PaddleStroke**
- Fixes https://github.com/FreeCAD/FreeCAD/issues/14966
- Also refactor a bit by adding finishDimensionCreation to reduce code copies.
- Fix https://github.com/FreeCAD/FreeCAD/issues/12033

### CODE REVIEW

**Comment by PaddleStroke** [comment]
Please consider merging this @chennes 
I say that because today I'm making a lot of conflicts with this PR while solving another bug. Which is annoying. Especially since this PR is just going to rot for one more week until next merge meeting where it will be like 'just a fix, let's merge it'.

**Comment by PaddleStroke** [comment]
Never mind I will add the fix to the other bug to this PR to avoid the conflicts.

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp [hunk 1]
```diff
@@ -1671,16 +1672,22 @@ class DrawSketchHandlerDimension : public DrawSketchHandler
...
         // Ask for the value of datum constraints
         ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher");
         bool show = hGrp->GetBool("ShowDialogOnDistanceConstraint", true);
         const std::vector<Sketcher::Constraint*>& ConStr = Obj->Constraints.getValues();
 
         bool commandHandledInEditDatum = false;
-        for (int i = numberOfConstraintsCreated - 1; i >= 0; i--) {
-            if (show && ConStr[ConStr.size() - 1 - i]->isDimensional() && ConStr[ConStr.size() - 1 - i]->isDriving) {
+        for (int i = cstrIndexes.size() - 1; i >= 0; --i) {
+            int index = cstrIndexes[i];
```
Comment (chennes) [inline src/Mod/Sketcher/Gui/CommandConstraints.cpp:1687]: Consider using either `#include <boost/range/adaptor/reversed.hpp>` in a range-based for loop, or the iterator-based version using `rbegin()` and `rend()`.

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp [hunk 2]
```diff
@@ -2708,14 +2667,54 @@ class DrawSketchHandlerDimension : public DrawSketchHandler
...
+    void addConstraintIndex()
+    {
+        cstrIndexes.push_back(Obj->Constraints.getValues().size() - 1);
+    }
+
+    bool hasBeenAborted()
+    {
+        // User can abort the command with Undo (ctrl-Z)
+        if (cstrIndexes.size() > 0) {
+            if (cstrIndexes[cstrIndexes.size() - 1] != Obj->Constraints.getValues().size() - 1 ) {
```
Comment (chennes) [inline src/Mod/Sketcher/Gui/CommandConstraints.cpp:2692]: Consider using `.back()`, e.g.
```suggestion
            if (cstrIndexes.back() != Obj->Constraints.getValues().size() - 1 ) {
```

Comment (PaddleStroke) [reply src/Mod/Sketcher/Gui/CommandConstraints.cpp:2692]: Oh yes back() is good.

**Comment by PaddleStroke** [comment]
Comments fixed

**Pull request merged**